File: .claude/skills/cernji-logging/assets/src/cernji-logging-ts/src/context.ts
/**
 * Correlation ID context management using AsyncLocalStorage
 */

import { AsyncLocalStorage } from 'async_hooks';
import { randomUUID } from 'crypto';

// AsyncLocalStorage to store correlation ID
const correlationIdStorage = new AsyncLocalStorage<string>();

/**
 * Get the current correlation ID from context
 */
export function getCorrelationId(): string | undefined {
  return correlationIdStorage.getStore();
}

/**
 * Set correlation ID and execute a function within that context
 *
 * @param correlationId - The correlation ID to set (generates new UUID if null)
 * @param fn - Async function to execute with the correlation ID in context
 * @returns The result of the function
 *
 * @example
 * ```typescript
 * await withCorrelationId('req-12345', async () => {
 *   logger.info('Processing request'); // Includes trace.id
 * });
 * ```
 */
export async function withCorrelationId<T>(
  correlationId: string | null,
  fn: () => Promise<T>
): Promise<T> {
  const id = correlationId || generateCorrelationId();
  return correlationIdStorage.run(id, fn);
}

/**
 * Generate a new correlation ID
 */
export function generateCorrelationId(): string {
  return randomUUID();
}

/**
 * Mixin function to inject correlation ID into log objects
 * Used by Pino to automatically add trace.id to all log records
 */
export function correlationIdMixin() {
  const correlationId = getCorrelationId();
  if (correlationId) {
    return {
      trace: {
        id: correlationId,
      },
    };
  }
  return {};
}

File: libs/cernji-logging-py/src/cernji_logging/context.py
"""Correlation ID context management using contextvars."""

import contextvars
from contextlib import contextmanager
from typing import Generator
from uuid import uuid4

# Context variable to store correlation ID
_correlation_id_var: contextvars.ContextVar[str | None] = contextvars.ContextVar(
    "correlation_id", default=None
)


def get_correlation_id() -> str | None:
    """Get the current correlation ID from context.

    Returns:
        The current correlation ID, or None if not set.
    """
    return _correlation_id_var.get()


def set_correlation_id(correlation_id: str) -> None:
    """Set the correlation ID in the current context.

    Args:
        correlation_id: The correlation ID to set.
    """
    _correlation_id_var.set(correlation_id)


def generate_correlation_id() -> str:
    """Generate a new correlation ID.

    Returns:
        A new UUID-based correlation ID.
    """
    return str(uuid4())


@contextmanager
def with_correlation_id(correlation_id: str | None = None) -> Generator[str, None, None]:
    """Context manager for setting correlation ID.

    Args:
        correlation_id: The correlation ID to use. If None, generates a new one.

    Yields:
        The correlation ID that was set.

    Example:
        ```python
        with with_correlation_id("req-12345") as cid:
            logger.info("Processing request")  # Includes correlation_id
        ```
    """
    if correlation_id is None:
        correlation_id = generate_correlation_id()

    token = _correlation_id_var.set(correlation_id)
    try:
        yield correlation_id
    finally:
        _correlation_id_var.reset(token)


def correlation_id_processor(logger, method_name, event_dict):
    """Structlog processor to inject correlation ID into log records.

    This processor is automatically added to the structlog configuration.
    """
    correlation_id = get_correlation_id()
    if correlation_id is not None:
        event_dict["trace.id"] = correlation_id
    return event_dict
